aip/general/0130.md

```diff
@@ -0,0 +1,77 @@
+| Batch resource methods             | Fetch or mutate multiple resources in a collection by name.                                                                                                                                                   | [AIP-231][], [AIP-233][], [AIP-234][], [AIP-235][] | may be used to optimize queries | automatable     | automatable    | automatable     |
+| **Custom Methods**                 |                                                                                                                                                                                                               |                                                    |                                 |                 |                |                 |
+| Custom collection fetch methods    | Fetch information across a collection that cannot be expressed via a standard method.                                                                                                                         | [AIP-136][]                                        | automatable                     | automatable     | automatable    | automatable     |
+| Custom collection mutation methods | Perform an imperative operation referencing a collection that may mutate one or more resources within that collection in fashion that cannot be easily achieved by standard methods (e.g. state transitions). | [AIP-136][]                                        | handwritten                     | automatable     | automatable    | automatable     |
+| Custom resource fetch methods      | Fetch information for a single resource that cannot be expressed via a standard method.                                                                                                                       | [AIP-136][]                                        | automatable                     | automatable     | automatable    | automatable     |
+| Custom resource mutation methods   | Perform an imperative operation on a resource that may mutate it in a way a standard method cannot (e.g. state transitions).                                                                                  | [AIP-136][]                                        | handwritten                     | automatable     | automatable    | automatable     |
+| **None of the above**              |                                                                                                                                                                                                               |                                                    |                                 |                 |                |                 |
```

Comment:
There are some categories of methods that are omitted:
* Aggregated list methods: automatable except for IaC, where they are not useful
* Snapshot/backup/commit and copy/restore/rollback custom methods are not really usable for IaC
* Resource rename methods are not usable for IaC
* Data-plane methods, including resourceless methods

Reply:
I was trying to avoid the "data-plane" term since I'm not sure that is descriptive enough as to the intent of the method, and since some things (e.g. snapshot) could be considered data-plane as well.

I can add the other ones though. I view the snapshot / backup / commit ones as more in the category of "mutation custom methods", so maybe it's worth having a whole category for that?

Reply:
re resourceless methods - I'm not convinced such a concept should exist. At least for GCP, everything should be done in the context of an org / folder / project. 

Although I guess categorically this refers to the more computational RPCs (e.g. speech to text)?

I could add it here, but we would need to clarify this with an AIP on that specifically I think.

Reply:
The management vs. data-plane distinction is important, as we've been discussing. It's mentioned in AIP 128 (https://google.aip.dev/128#guidance) but I don't think it's discussed in AIP 136. We probably need a dedicated AIP on that topic, so I don't think we need a lot of detail here, but this document seems to assume it's talking about only the management plane without saying so.

Resourceless methods are called stateless methods in AIP 136: https://google.aip.dev/136#stateless-methods. Resources often have user data attached to them, like a storage object, database table, VM memory, etc. and provide data-plane methods to write and read that data. Resourceless/stateless methods operate on data attached to other resources, such as storage objects. I believe the concept does make sense. They are pure functions, essentially. How we map the concept to an API pattern is another issue. Such a method may have an explicit parent for authorization, policy, billing, and regionalization purposes. If we want to require an explicit resource even in the case that there is no persistent state associated with the service, it would be a singleton, however we decide to represent those, so I'd suggest we tackle singleton first, then stateless methods.

Reply:
> The management vs. data-plane distinction is important, as we've been discussing. It's mentioned in AIP 128 (https://google.aip.dev/128#guidance) but I don't think it's discussed in AIP 136. We probably need a dedicated AIP on that topic, so I don't think we need a lot of detail here, but this document seems to assume it's talking about only the management plane without saying so.

I do agree we need to tackle management plane vs data plane, but I'm not sure about the right way to express that in this document. resource-oriented CRUDL also exists in data plane as well. 

If we do tackle the discussion of the plane, I think it makes sense to do so in this AIP.

> Resourceless methods are called stateless methods in AIP 136: https://google.aip.dev/136#stateless-methods.

Ack, I'll add a line for stateless methods. I think it's a good point that how we categorize the method and how the method is expressed via HTTP are two separate concerns. 

I don't know if I view it as a singleton though: operations on a singleton mutate data, while stateless methods do not.